examples/paper/03-dependent-binding.makam

```diff
@@ -1,90 +1,216 @@
+   λProlog is simple: before performing unification at the term-level, we perform unification at
+   the type level first, therefore further determining any uninstantiated type variables. Therefore,
+   when we check to see whether the current goal matches the premise of a rule, type unification
+   can force us to distinguish between different types. Based on these, the standard example
+   of `map` for vectors is as follows:
+*)
+
+vmap : [A B] (A -> B -> prop) -> vector N A -> vector N B -> prop.
+vmap P vnil vnil.
+vmap P (vcons X XS) (vcons Y YS) :- P X Y, vmap P XS YS.
+
+(* The notation [N] in the type of `vmap` means that the type argument `N` is ad-hoc/not-parametric.
+   Non-specified type arguments are parametric by default, so as to match standard practice in
+   languages like ML and Haskell, and to catch type errors that allowing unqualified ad-hoc
+   polymorphism would permit. For example, consider the following erroneous definition for `fold`,
+   where the arguments for `P` in the `cons` case are flipped.
+
+foldl : (B -> A -> B -> prop) -> B -> list A -> B -> prop.
+foldl P S nil S.
+foldl P S (cons HD TL) S'' <- P HD S S', foldl P S' TL S''.
+
+   If ad-hoc polymorphism is allowed for `A` and `B`, this is a well-typed definition. However, the
+   erroneous call to `P` forces the types `A` and `B` to be unified, and therefore the `fold`
+   predicate is unnecessarily restricted to only work when the two types are the same. Having to
+   specify ad-hoc polymorphism explicitly helps us avoid such mistakes. *)
+
+(* Though this support for ad-hoc polymorphism was part of the original λProlog design, we have not
+   found extensive coverage of its implications in the literature. Furthermore, it is not supported
+   well by standard implementations of λProlog (like Teyjus), which was one of the reasons that
+   prompted us to work on Makam. 
+*)
+
+(* Armed with GADTs of this form, we can now introduce dependently-typed binding forms, where the
+   number of variables that are being bound is reflected in the type. One way to do this is the
+   following: 
+*)
+
+(* dbind A N B: dependently-typed binding of N A's into a B; N will be instantiated with `natZ`
+   and `natS` as above. *)
+
+dbind : type -> type -> type -> type. 
+
+dbindbase : B -> dbind A natZ B.
+dbindnext : (A -> dbind A N B) -> dbind A (natS N) B.
 
-(* dbind A T B: dependently-typed binding of many A's into a B; T is a tuple-type representing the type of substitutions. *)
+(* Another possibility, avoiding the need for introducing type-level natural numbers, is to
+   use a more standard type as the dependent parameter: the type of tuples that would serve
+   as substitutions for the introduced variables. The type would then become:
+*)
 
 dbind : type -> type -> type -> type.
 
-%extend dbind.
+dbindbase : B -> dbind A unit B.
+dbindnext : (A -> dbind A T B) -> dbind A (A * T) B.
+
```

Comment:
@achlipala One thing we need to decide here is which one to go with -- the vector-like encoding or the tuple-based one.

The good thing about vectors is that they're well-understood from the dependent types literature, and also make for quite pleasant higher-order predicates that are entirely equivalent to lists and bindmany. As a result, we could elide a lot of the details in the paper, giving only a few examples.

The bad thing about them is the clunky representation for natural numbers: Makam was not designed to do actual dependent types, and adding the option for introducing new kinds other than `type`, such as `nat`, would be both time-consuming, and would also necessitate us talking more about the actual Makam type system. I feel like if we keep it at the subset of Fω that it is now, the hand-wavy explanation at the top of this file is enough. Also, the other problem is that they don't generalize as well: the tuple representation is very easy to generalize to variables of different sorts (just get rid of the first type argument), which could be useful in later developments -- it will play nicely in terms of the expression problem. I wouldn't want to introduce heterogeneous lists etc. instead.

The problem with tuples, of course, is that the higher-order predicates etc. don't look as nice, and require quite a bit of ad-hoc polymorphism: the whole type of the tuple has to stay abstract and be pattern-matched over, whereas with vectors we only do ad-hoc polymorphism for the 'dependent' argument. I've introduced the `subst` type below to mitigate the issue somewhat.

I quite like the fact that the very basic type system of Makam, together with the ad-hoc polymorphism that comes for free in the λProlog world, yields such a nice (I think) and easy to work with notion of dependency that is useful for our purposes. This too points towards using tuples, as it does not feel like something is 'missing', as happens when we introduce hacks like `natZ` and `natS`.